sql: migrate the tracing flags to cluster settings.

knz opened the pull request:
Prior to this patch CockroachDB supported enabling SQL tracing
via environment variables; this patch replaces this mechanism
by cluster settings.

| Env var (old)                  | Setting (new)                        |
|--------------------------------|--------------------------------------|
| COCKROACH_TRACE_SQL            | `sql.trace.txn.threshold`            |
| COCKROACH_ENABLE_SQL_EVENT_LOG | `sql.trace.session.eventlog.enabled` |


In addition, this patch also introduces a new setting
`sql.log.statements.execute.enabled` which, when set to true, causes
executed statements to be copied to the logging output. This mechanism
is more lightweight than the tracing options above since it does not
need to accumulate state in the session.

Note that `--vmodule=executor=2` still forces statement logging,
regardless of the setting's current value.

Fixes #13851.

knz committed:
sql: migrate the tracing flags to cluster settings.

Prior to this patch CockroachDB supported enabling SQL tracing
via environment variables; this patch replaces this mechanism
by cluster settings.

| Env var (old)                  | Setting (new)                        |
|--------------------------------|--------------------------------------|
| COCKROACH_TRACE_SQL            | `sql.trace.txn.threshold`            |
| COCKROACH_ENABLE_SQL_EVENT_LOG | `sql.trace.session.eventlog.enabled` |

In addition, this patch also introduces a new setting
`sql.log.statements.execute.enabled` which, when set to true, causes
executed statements to be copied to the logging output. This mechanism
is more lightweight than the tracing options above since it does not
need to accumulate state in the session.

Note that `--vmodule=executor=2` still forces statement logging,
regardless of the setting's current value.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15190)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 570 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-Ki80qMsobejZA-L8q52:-Ki80qMsobejZA-L8q53:bkxs3vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b86c942906242c02597af96373d84ce8cc11b5a7/pkg/sql/executor.go#L570)):*
> ```Go
> 	txnState := &session.TxnState
> 
> 	if log.V(2) || atomic.LoadInt32(&logStatementsExecuteEnabled) != 0 {
> ```

The repeated use of `atomic.LoadInt32` is awkward. Perhaps there should be a `logStatementsExecuteEnabled()` function.

---

*[pkg/sql/session.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-Ki80KjXcB6BsO_mJYzB:-Ki80KjXcB6BsO_mJYzC:btxa059) ([raw file](https://github.com/cockroachdb/cockroach/blob/b86c942906242c02597af96373d84ce8cc11b5a7/pkg/sql/session.go#L70)):*
> ```Go
> // own tracer and log, in the hope that we'll catch #7881 and dump the
> // current trace for debugging.
> var debugTrace7881Enabled int32
> ```

@andreimatei Do we still need this? And if we do, should it be in the `system.settings` table?

---

*[pkg/sql/session.go, line 694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-Ki81E4TtAGwUrNmKVn2:-Ki81E4TtAGwUrNmKVn3:bejqqhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b86c942906242c02597af96373d84ce8cc11b5a7/pkg/sql/session.go#L694)):*
> ```Go
> 	ts.sp = nil
> 	if ts.trace != nil {
> 		durThreshold := time.Duration(atomic.LoadInt64(&traceTxnThreshold))
> ```

Mildly strange that `traceTxnThreshold` can change during the lifetime of the query, but I can't see that actually causing a problem in practice. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-:-Ki81Nh-26jcDtTltByN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz committed:
settings: further improve the settings accessors.

Prior to this patch, the accessor function returned by the
`Register*Setting` functions would call the getXXX() accessors, which
in turn demand two dictionary lookups and a mutex lock. This was
already initially foreseen to be an issue, which is why the
`RegisterCallback()` function was also provided so that client code
can use that to update local variables that can then be read without
the overhead.

However in practice it was revealed that the trade-off was not very
convenient. The alternative with `RegisterCallback()` requires atomic
accesses with `atomic.Load`/`atomic.Store` calls peppered throughout,
to avoid data races since settings can be updated asynchronously.
This in turn becomes very inconvenient for settings with type "bool",
"float" and "duration" because these types do not fit naturally in the
only two types int32/int64 supported by the `atomic` package.

To make the interface more convenient, this patch changes the return
type of the `Register*Setting` functions to become an object that
already supports the atomic access without overhead natively, except
for settings of type `string` where there is no atomic access possible
and thus the overhead of a mutex lock is unavoidable.

knz commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 570 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-Ki80qMsobejZA-L8q52:-KiAP_KbqySXZcD7nagr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b86c942906242c02597af96373d84ce8cc11b5a7/pkg/sql/executor.go#L570)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The repeated use of `atomic.LoadInt32` is awkward. Perhaps there should be a `logStatementsExecuteEnabled()` function.
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-Ki80KjXcB6BsO_mJYzB:-KiAPb0kiXFG1OZngj83:b-l7dubw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b86c942906242c02597af96373d84ce8cc11b5a7/pkg/sql/session.go#L70)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

@andreimatei Do we still need this? And if we do, should it be in the `system.settings` table?
</blockquote></details>

We discussed this and agreed this should be enabled only for tests, so an env var would be sufficient. Reverted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190)*
<!-- Sent from Reviewable.io -->

dt commented:
Ha! -- your sync/atomic additions are almost exactly what I have locally. 

---

Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/settings/accessors.go, line 96 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-KiARsAbHYLN6btiJPjV:-KiARsAbHYLN6btiJPjW:b-iwn7j4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea002087428f6a9017782cc87b09995c02258d2d/pkg/settings/accessors.go#L96)):*
> atomic.LoadInt32(&frozen) > 0 {
> 		panic(fmt.Sprintf("registration must occur before server start: %s", key))
> 	}
> 	if _, ok := registry[key]; ok {
> 		panic(fmt.Sprintf("setting already defined: %s", key))
> 	}
could DRY this between helpers.

---

*[pkg/settings/accessors.go, line 102 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-KiASD5QkixQLqB3n1nS:-KiASD5QkixQLqB3n1nT:box4edr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea002087428f6a9017782cc87b09995c02258d2d/pkg/settings/accessors.go#L102)):*
> ```Go
> 	v := int32(0)
> 	setting := BoolSetting{v: &v}
> 	RegisterCallback(func() {
> ```

This works, but registering individual callbacks for all of them might be more expensive than just putting the value pointer in the registry and then iterating over it in afterApply?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-:-KiARwqE86aO2l3Mnha8:buuao1r)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/settings/accessors.go, line 96 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-KiARsAbHYLN6btiJPjV:-KiAU6hIMjCmXSi6W26P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea002087428f6a9017782cc87b09995c02258d2d/pkg/settings/accessors.go#L96)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

> atomic.LoadInt32(&frozen) > 0 {
> 		panic(fmt.Sprintf("registration must occur before server start: %s", key))
> 	}
> 	if _, ok := registry[key]; ok {
> 		panic(fmt.Sprintf("setting already defined: %s", key))
> 	}
could DRY this between helpers.
</blockquote></details>

Done.

---

*[pkg/settings/accessors.go, line 102 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-KiASD5QkixQLqB3n1nS:-KiAUBQCyN8ynneHPxDG:bz5py0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea002087428f6a9017782cc87b09995c02258d2d/pkg/settings/accessors.go#L102)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

This works, but registering individual callbacks for all of them might be more expensive than just putting the value pointer in the registry and then iterating over it in afterApply?
</blockquote></details>

Good idea! Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I didn't look at the `settings` changes, but the usage in `sql` looks nice now. 

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-:-KiAYngsgaL8XgFTYp37:b-sjp8kg)*
<!-- Sent from Reviewable.io -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/settings/accessors.go, line 102 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-KiASD5QkixQLqB3n1nS:-KiAaiMcgs_vOOy6Evbc:b-a9v1v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea002087428f6a9017782cc87b09995c02258d2d/pkg/settings/accessors.go#L102)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Good idea! Done.
</blockquote></details>

heh, this still isn't _quite_ what I meant : I more meant we didn't need to create closures at all and could just put values in there, but that can be a followup e.g. #15200

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15190#-:-KiAcAkk3oYK9Q4wlO6q:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks! Yes, I'll improve this incrementally as we learn more how to use it.
